presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/WindowMatcher.java

```diff
@@ -18,58 +18,49 @@
 import com.facebook.presto.metadata.Metadata;
 import com.facebook.presto.sql.planner.plan.PlanNode;
 import com.facebook.presto.sql.planner.plan.WindowNode;
-import com.facebook.presto.sql.tree.FunctionCall;
-import com.google.common.collect.ImmutableList;
-
-import java.util.LinkedList;
-import java.util.List;
-import java.util.stream.Collectors;
 
 import static com.google.common.base.MoreObjects.toStringHelper;
-import static java.util.Objects.requireNonNull;
+import static com.google.common.base.Preconditions.checkState;
 
 final class WindowMatcher
         implements Matcher
 {
-    private final List<FunctionCall> functionCalls;
+    private final ExpectedValueProvider<WindowNode.Specification> specification;
 
-    WindowMatcher(List<FunctionCall> functionCalls)
+    WindowMatcher(
+            ExpectedValueProvider<WindowNode.Specification> specification)
     {
-        this.functionCalls = ImmutableList.copyOf(requireNonNull(functionCalls, "functionCalls is null"));
+        this.specification = specification;
     }
 
     @Override
-    public boolean matches(PlanNode node, Session session, Metadata metadata, ExpressionAliases expressionAliases)
+    public boolean downMatches(PlanNode node)
     {
-        if (!(node instanceof WindowNode)) {
-            return false;
-        }
-
-        WindowNode windowNode = (WindowNode) node;
-        LinkedList<FunctionCall> actualCalls = windowNode.getWindowFunctions().values().stream()
-                .map(WindowNode.Function::getFunctionCall)
-                .collect(Collectors.toCollection(LinkedList::new));
+        return node instanceof WindowNode;
+    }
 
-        if (actualCalls.size() != functionCalls.size()) {
-            return false;
-        }
+    @Override
+    public boolean upMatches(PlanNode node, Session session, Metadata metadata, ExpressionAliases expressionAliases)
+    {
+        checkState(downMatches(node), "DSL framework error: downMatches returned false in upMatches in %s", this.getClass().getName());
 
-        for (FunctionCall expectedCall : functionCalls) {
-            if (!actualCalls.remove(expectedCall)) {
-                // Found an expectedCall not in expectedCalls.
-                return false;
-            }
-        }
+        WindowNode windowNode = (WindowNode) node;
 
-        // expectedCalls was missing something in actualCalls.
-        return actualCalls.isEmpty();
+        /*
+         * Window functions are tested using an Alias matcher, not here.
```

Comment:
What does this mean? You test the specification here, what else is to be tested?